pkg/chunk/cache/cache.go

```diff
@@ -37,19 +47,52 @@ type Config struct {
 	Cache Cache
 }
 
+// StoreConfig for configuring data storage systems
+type StoreConfig struct {
+	Type           string        `yaml:"type"`
+	Host           string        `yaml:"host,omitempty"`
+	Service        string        `yaml:"service,omitempty"`
+	Timeout        time.Duration `yaml:"timeout,omitempty"`
+	MaxIdleConns   int           `yaml:"max_idle_conns,omitempty"`
+	MaxActiveConns int           `yaml:"max_active_conns,omitempty"`
+	UpdateInterval time.Duration `yaml:"update_interval,omitempty"`
+	Expiration     time.Duration `yaml:"expiration,omitempty"`
+	BatchSize      int           `yaml:"batch_size,omitempty"`
+	Parallelism    int           `yaml:"parallelism,omitempty"`
+	ConsistentHash bool          `yaml:"consistent_hash,omitempty"`
+}
+
 // RegisterFlagsWithPrefix adds the flags required to config this to the given FlagSet
 func (cfg *Config) RegisterFlagsWithPrefix(prefix string, description string, f *flag.FlagSet) {
 	cfg.Background.RegisterFlagsWithPrefix(prefix, description, f)
+	cfg.Store.RegisterFlagsWithPrefix(prefix, description, f)
+	cfg.Fifocache.RegisterFlagsWithPrefix(prefix, description, f)
+
+	// backward compatibility for memcached settings (to be deprecated)
 	cfg.Memcache.RegisterFlagsWithPrefix(prefix, description, f)
 	cfg.MemcacheClient.RegisterFlagsWithPrefix(prefix, description, f)
-	cfg.Fifocache.RegisterFlagsWithPrefix(prefix, description, f)
 
 	f.BoolVar(&cfg.EnableFifoCache, prefix+"cache.enable-fifocache", false, description+"Enable in-memory cache.")
 	f.DurationVar(&cfg.DefaultValidity, prefix+"default-validity", 0, description+"The default validity of entries for caches unless overridden.")
 
 	cfg.Prefix = prefix
 }
 
+// RegisterFlagsWithPrefix adds the flags required to config this to the given FlagSet
+func (cfg *StoreConfig) RegisterFlagsWithPrefix(prefix, description string, f *flag.FlagSet) {
+	f.StringVar(&cfg.Type, prefix+"cache.store.type", "", description+"Type of storage cache.")
+	f.DurationVar(&cfg.Expiration, prefix+"cache.store.expiration", 0, description+"How long keys stay in the cache.")
```

Comment:
I'm not sure this is the best way to do things tbh. Some of the options here are for one store and then others are for others stores.

I'd rather have:
```
StoreConfig {
   Type string
   Redis RedisConfig
   Memcached MemcachedConfig
   NewCacheType  NewCacheTypeConfig 
}
```

That'll make things easier and we'll clearly know what to set for what.

Further the flags would be:
`cache.type`, `cache.memcache.*`, `cache.redis.*`.

Reply:
Also makes things like `NewMemcachedClient()` take only `Memcached` parameters.